File: python/docs/examples/codespace/code_execution_example.py
#!/usr/bin/env python3
"""
AgentBay SDK - CodeSpace Example

This example demonstrates how to use AgentBay SDK code execution features:
- Python code execution
- JavaScript code execution
- File operations in code environment
- Command execution in code environment
"""

import os
from agentbay import AgentBay
from agentbay.session_params import CreateSessionParams


def main():
    """Main function demonstrating code execution features"""
    print("🚀 AgentBay CodeSpace Example")

    # Initialize AgentBay client
    api_key = os.getenv("AGENTBAY_API_KEY")
    if not api_key:
        print("❌ Error: AGENTBAY_API_KEY environment variable not set")
        return

    agent_bay = AgentBay(api_key=api_key)

    # Create session with code_latest image
    print("\n📱 Creating session with code_latest image...")
    params = CreateSessionParams(image_id="code_latest")
    session_result = agent_bay.create(params)

    if not session_result.success:
        print(f"❌ Session creation failed: {session_result.error_message}")
        return

    session = session_result.session
    print(f"✅ Session created successfully: {session.session_id}")

    try:
        # ===== PYTHON CODE EXECUTION =====
        print("\n===== PYTHON CODE EXECUTION =====")

        python_code = """
import sys
import os
import json
from datetime import datetime

# System information
system_info = {
    "python_version": sys.version,
    "current_directory": os.getcwd(),
    "timestamp": datetime.now().isoformat(),
    "environment_vars": len(os.environ)
}

print("Python code execution successful!")
print(f"System info: {json.dumps(system_info, indent=2)}")

# Simple calculation
numbers = list(range(1, 11))
total = sum(numbers)
print(f"Sum of 1 to 10: {total}")
"""

        print("🔄 Executing Python code...")
        result = session.code.run_code(python_code, "python")
        if result.success:
            print("✅ Python code executed successfully:")
            print(result.result)
        else:
            print(f"❌ Python code execution failed: {result.error_message}")

        # ===== JAVASCRIPT CODE EXECUTION =====
        print("\n===== JAVASCRIPT CODE EXECUTION =====")

        js_code = """
console.log("JavaScript code execution successful!");

// Get system information
const os = require('os');
const systemInfo = {
    platform: os.platform(),
    arch: os.arch(),
    nodeVersion: process.version,
    memory: Math.round(os.totalmem() / 1024 / 1024) + ' MB'
};

console.log("System info:", JSON.stringify(systemInfo, null, 2));

// Array operations
const numbers = [1, 2, 3, 4, 5];
const doubled = numbers.map(n => n * 2);
console.log("Original array:", numbers);
console.log("Doubled array:", doubled);
"""

        print("🔄 Executing JavaScript code...")
        result = session.code.run_code(js_code, "javascript")
        if result.success:
            print("✅ JavaScript code executed successfully:")
            print(result.result)
        else:
            print(f"❌ JavaScript code execution failed: {result.error_message}")

        # ===== FILE OPERATIONS =====
        print("\n===== FILE OPERATIONS =====")

        print("🔄 Testing file operations...")
        test_content = "Hello from AgentBay code execution!"
        test_file_path = "/tmp/test_code.txt"

        write_result = session.file_system.write_file(test_file_path, test_content)
        if write_result.success:
            print("✅ File written successfully")

            read_result = session.file_system.read_file(test_file_path)
            if read_result.success:
                print(f"✅ File content: {read_result.content}")
            else:
                print(f"❌ File read failed: {read_result.error_message}")
        else:
            print(f"❌ File write failed: {write_result.error_message}")

        # ===== COMMAND EXECUTION =====
        print("\n===== COMMAND EXECUTION =====")

        commands = [
            "whoami",
            "pwd",
            "python3 --version",
            "node --version",
            "ls -la /tmp"
        ]

        for cmd in commands:
            print(f"\n🔄 Executing command: {cmd}")
            result = session.command.execute_command(cmd)

            if result.success:
                print(f"✅ Output: {result.output.strip()}")
            else:
                print(f"❌ Command failed: {result.error_message}")

    finally:
        # Clean up session
        print(f"\n🧹 Cleaning up session: {session.session_id}")
        delete_result = agent_bay.delete(session)
        if delete_result.success:
            print("✅ Session deleted successfully")
        else:
            print(f"❌ Failed to delete session: {delete_result.error_message}")


if __name__ == "__main__":
    main()

File: java/agentbay/src/main/java/com/aliyun/agentbay/examples/CodeExecutionExample.java
package com.aliyun.agentbay.examples;

import com.aliyun.agentbay.AgentBay;
import com.aliyun.agentbay.Config;
import com.aliyun.agentbay.filesystem.FileSystem;
import com.aliyun.agentbay.model.*;
import com.aliyun.agentbay.exception.AgentBayException;
import com.aliyun.agentbay.exception.SessionException;
import com.aliyun.agentbay.model.code.EnhancedCodeExecutionResult;
import com.aliyun.agentbay.session.CreateSessionParams;
import com.aliyun.agentbay.session.Session;;

/**
 * Example demonstrating code execution functionality in AgentBay Java SDK
 */
public class CodeExecutionExample {

    public static void main(String[] args) {
        try {
            // Get API Key from environment variable
            String apiKey = System.getenv("AGENTBAY_API_KEY");
            if (apiKey == null || apiKey.trim().isEmpty()) {
                System.err.println("Error: AGENTBAY_API_KEY environment variable not set");
                return;
            }

            // Get configuration from environment variables or use defaults
            String region = getEnvOrDefault("AGENTBAY_REGION", "us-east-1");
            String endpoint = getEnvOrDefault("AGENTBAY_ENDPOINT", "agentbay.us-east-1.aliyuncs.com");
            int timeout = Integer.parseInt(getEnvOrDefault("AGENTBAY_TIMEOUT", "60000"));
            String imageId = getEnvOrDefault("AGENTBAY_IMAGE_ID", "imgc-0aae4rxtt0yuix7oh");

            // Print configuration
            System.out.println("========== Configuration ==========");
            System.out.println("Region: " + region);
            System.out.println("Endpoint: " + endpoint);
            System.out.println("Timeout: " + timeout + "ms");
            System.out.println("Image ID: " + imageId);
            System.out.println("===================================\n");

            // Create AgentBay client
            System.out.println("Creating AgentBay client...");
            AgentBay agentBay = new AgentBay(apiKey, new Config(region, endpoint, timeout));


            // Create a session
            System.out.println("Creating session...");
            CreateSessionParams params = new CreateSessionParams();
            params.setImageId(imageId);
            long startTime = System.currentTimeMillis();
            SessionResult sessionResult = agentBay.create(params);
            long duration = System.currentTimeMillis() - startTime;
            System.out.println("⏱️  Create session finished in " + duration + "ms");

            if (!sessionResult.isSuccess()) {
                System.err.println("Failed to create session: " + sessionResult.getErrorMessage());
                return;
            }

            Session session = sessionResult.getSession();

            System.out.println("✅ Session created successfully!");
            System.out.println("   Session ID: " + session.getSessionId());
            System.out.println("   Request ID: " + sessionResult.getRequestId());

            FileSystem fs = session.getFileSystem();
//
//            // ===== BASIC FILE OPERATIONS =====
//            System.out.println("\n===== BASIC FILE OPERATIONS =====");
//
            // Example 1: Write a simple file
            System.out.println("\nExample 1: Writing a simple file...");
            String testContent = "This is a test file content.\nIt has multiple lines.\nThis is the third line.";
            String testFilePath = "/tmp/test_file.txt";

            long writeStartTime = System.currentTimeMillis();
            BoolResult result1 = fs.writeFile(testFilePath, testContent, "overwrite");
            long writeDuration = System.currentTimeMillis() - writeStartTime;
            System.out.println("File write successful: " + result1.isSuccess());
            System.out.println("⏱️  Write file finished in " + writeDuration + "ms");
            if (!result1.isSuccess()) {
                System.out.println("Error: " + result1.getErrorMessage());
            }
            System.out.println("Request ID: " + result1.getRequestId());


            // Example 1: Execute Python code
            System.out.println("\n🐍 Executing Python code...");
            String pythonCode = "print(\"hello world\")";
            long pythonStartTime = System.currentTimeMillis();
            EnhancedCodeExecutionResult pythonResult = session.getCode().runCode(pythonCode, "python");
            long pythonDuration = System.currentTimeMillis() - pythonStartTime;

            if (pythonResult.isSuccess()) {
                System.out.println("✅ Python code executed successfully!");
                System.out.println("⏱️  Python execution finished in " + pythonDuration + "ms");
                System.out.println("   Output: " + pythonResult.getResult());
                System.out.println("   Request ID: " + pythonResult.getRequestId());
            } else {
                System.err.println("❌ Failed to execute Python code: " + pythonResult.getErrorMessage());
                System.out.println("⏱️  Python execution finished in " + pythonDuration + "ms");
            }

            // Example 2: Execute JavaScript code
            System.out.println("\n🟨 Executing JavaScript code...");
            String jsCode = "console.log('Hello from JavaScript!');\nconst result = 5 * 4;\nconsole.log(`5 * 4 = ${result}`);";
            long jsStartTime = System.currentTimeMillis();
            EnhancedCodeExecutionResult jsResult = session.getCode().runCode(jsCode, "javascript");
            long jsDuration = System.currentTimeMillis() - jsStartTime;

            if (jsResult.isSuccess()) {
                System.out.println("✅ JavaScript code executed successfully!");
                System.out.println("⏱️  JavaScript execution finished in " + jsDuration + "ms");
                System.out.println("   Output: " + jsResult.getResult());
                System.out.println("   Request ID: " + jsResult.getRequestId());
            } else {
                System.err.println("❌ Failed to execute JavaScript code: " + jsResult.getErrorMessage());
                System.out.println("⏱️  JavaScript execution finished in " + jsDuration + "ms");
            }

            // Clean up - delete the session
            DeleteResult deleteResult = session.delete();
            if (!deleteResult.isSuccess()) {
                System.err.println("Failed to delete session: " + deleteResult.getErrorMessage());
            }
            System.out.println("   Session deleted successfully!");


        } catch (SessionException e) {
            System.err.println("❌ Session error: " + e.getMessage());
            e.printStackTrace();
        } catch (AgentBayException e) {
            System.err.println("❌ AgentBay error: " + e.getMessage());
            e.printStackTrace();
        } catch (Exception e) {
            System.err.println("❌ Unexpected error: " + e.getMessage());
            e.printStackTrace();
        }
    }

    /**
     * Helper method to get environment variable or default value
     */
    private static String getEnvOrDefault(String envName, String defaultValue) {
        String value = System.getenv(envName);
        return (value != null && !value.trim().isEmpty()) ? value : defaultValue;
    }
}